x-pack/plugins/data_visualizer/public/application/index_data_visualizer/embeddables/field_stats/field_stats_factory.tsx

```diff
@@ -0,0 +1,379 @@
+    defaultMessage: 'Error applying filter',
+  }),
+};
+
+export const getDependencies = async (
+  getStartServices: StartServicesAccessor<
+    DataVisualizerStartDependencies,
+    DataVisualizerPluginStart
+  >
+) => {
+  const [
+    { http, uiSettings, notifications, ...startServices },
+    { lens, data, usageCollection, fieldFormats },
+  ] = await getStartServices();
+
+  return {
+    http,
+    uiSettings,
+    data,
+    notifications,
+    lens,
+    usageCollection,
+    fieldFormats,
+    ...startServices,
+  };
+};
+
+export const getFieldStatsChartEmbeddableFactory = (
+  getStartServices: StartServicesAccessor<
+    DataVisualizerStartDependencies,
+    DataVisualizerPluginStart
+  >
+) => {
+  const factory: ReactEmbeddableFactory<
+    FieldStatisticsTableEmbeddableState,
+    FieldStatisticsTableEmbeddableState,
+    FieldStatisticsTableEmbeddableApi
+  > = {
+    type: FIELD_STATS_EMBEDDABLE_TYPE,
+    deserializeState: (state) => {
+      const serializedState = cloneDeep(state.rawState);
+      // inject the reference
+      const dataViewIdRef = state.references?.find(
+        (ref) => ref.name === FIELD_STATS_DATA_VIEW_REF_NAME
+      );
+      // if the serializedState already contains a dataViewId, we don't want to overwrite it. (Unsaved state can cause this)
+      if (dataViewIdRef && serializedState && !serializedState.dataViewId) {
+        serializedState.dataViewId = dataViewIdRef?.id;
+      }
+      return serializedState;
+    },
+    buildEmbeddable: async (state, buildApi, uuid, parentApi) => {
+      const [coreStart, pluginStart] = await getStartServices();
+
+      const { http, uiSettings, notifications, ...startServices } = coreStart;
+      const { lens, data, usageCollection, fieldFormats } = pluginStart;
+
+      const deps = {
+        http,
+        uiSettings,
+        data,
+        notifications,
+        lens,
+        usageCollection,
+        fieldFormats,
+        ...startServices,
+      };
+      const {
+        api: timeRangeApi,
+        comparators: timeRangeComparators,
+        serialize: serializeTimeRange,
+      } = initializeTimeRange(state);
+
+      const { titlesApi, titleComparators, serializeTitles } = initializeTitles(state);
+
+      const {
+        fieldStatsControlsApi,
+        dataLoadingApi,
+        fieldStatsControlsComparators,
+        serializeFieldStatsChartState,
+        onFieldStatsTableDestroy,
+        resetData$,
+      } = initializeFieldStatsControls(state);
+      const { onError } = dataLoadingApi;
+
+      const defaultDataViewId = await deps.data.dataViews.getDefaultId();
+      const validDataViewId: string =
+        isDefined(state.dataViewId) && state.dataViewId !== ''
+          ? state.dataViewId
+          : defaultDataViewId ?? '';
+      let initialDataView: DataView[] | undefined;
+      try {
+        const dataView = await deps.data.dataViews.get(validDataViewId);
```

Comment:
In the catch block, there is an if statement for `isESQLQuery` that sets data view to adHocDataView. Why is this in the catch block? Shouldn't `const dataView = await deps.data.dataViews.get(validDataViewId);` only execute when its not an ESQL query?

Reply:
Because it's possible that the ES|QL panel was using an adhoc dataview previously, and when the page refreshes, it doesn't have that adhoc dataview anymore. For ES|QL, I create adhoc dataview based on the ES|QL query mainly so that dashboard filters know what data views to use. Without this, dashboard filters would look like this:

<img width="675" alt="image" src="https://github.com/elastic/kibana/assets/43350163/fef4b279-5d72-46f2-9c9b-e344b1237b9c">

Not sure what's a good way around this. If you have a recommended solution, please let me know.

Reply:
How about this
```
try {
    const dataView = isESQLQuery(state.query)
       ? await getESQLAdHocDataview(
            getIndexPatternFromESQLQuery(state.query.esql),
            deps.data.dataViews
          )
       : await deps.data.dataViews.get(validDataViewId);
        initialDataView = [dataView];
      }

```

[getESQLAdHocDataview](https://github.com/elastic/kibana/blob/main/packages/kbn-esql-utils/src/utils/get_esql_adhoc_dataview.ts#L37) uses a hash of the indexPattern for the id so the id will be the same for the same indexPattern every time getESQLAdHocDataview is called.

Reply:
Updated here https://github.com/elastic/kibana/pull/184030/commits/d36df76b45da69098e9ffb27f04fb4365acb45ec